src/test/java/com/maxmind/db/ReaderTest.java

```diff
@@ -620,6 +621,17 @@ public void testDecodeWrongTypeWithWrongArguments() throws IOException {
         assertThat(ex.getMessage(), containsString("Error getting record for IP"));
     }
 
+    @Test
+    public void testDecodeWithDataTypeMismatchInModel() throws IOException {
+        this.testReader = new Reader(getFile("GeoIP2-City-Test.mmdb"));
+        DeserializationException ex = assertThrows(DeserializationException.class,
+                () -> this.testReader.get(InetAddress.getByName("2.125.160.216"),
+                        TestDataTypeMismatchInModel.class));
+        assertThat(ex.getMessage(), containsString("Error getting record for IP"));
+        assertThat(ex.getMessage(), containsString("Error creating map entry for"));
+        assertThat(ex.getCause().getCause().getClass(), equalTo(ClassCastException.class));
```

Comment:
A utility that I find helpful for this sort of thing is ExceptionUtils.getRootCause(); this is part of the Apache commons library. In this case I've not used ExceptionUtils as to not bloat the repo with unnecessary libraries.